paimon-core/src/main/java/org/apache/paimon/table/source/InnerTableScanImpl.java

```diff
@@ -68,9 +81,96 @@ public TableScan.Plan plan() {
         if (hasNext) {
             hasNext = false;
             StartingScanner.Result result = startingScanner.scan(snapshotReader);
-            return DataFilePlan.fromResult(result);
+            StartingScanner.Result limitedResult = applyPushDownLimit(result);
+            return DataFilePlan.fromResult(limitedResult);
         } else {
             throw new EndOfScanException();
         }
     }
+
+    private StartingScanner.Result applyPushDownLimit(StartingScanner.Result result) {
+        if (pushDownLimit != null && result instanceof StartingScanner.ScannedResult) {
+            long scannedRowCount = 0;
+            SnapshotReader.Plan plan = ((StartingScanner.ScannedResult) result).plan();
+            List<DataSplit> splits = plan.dataSplits();
+            List<DataSplit> limitedSplits = new ArrayList<>();
+            for (int i = 0; i < splits.size(); i++) {
+                if (scannedRowCount >= pushDownLimit) {
+                    break;
+                }
+
+                DataSplit split = splits.get(i);
+                long splitRowCount = getRowCountForSplit(split);
+                if (scannedRowCount + splitRowCount <= pushDownLimit) {
+                    limitedSplits.add(split);
+                    scannedRowCount += splitRowCount;
+                } else {
+                    DataSplit newSplit = composeDataSplit(split, pushDownLimit - scannedRowCount);
```

Comment:
I think we don't need to introduce `composeDataSplit`.

The reason we introduce limit pushdown is to reduce split number. We don't need to reduce files in split.